BaseApp/COD3041.TXT

```diff
@@ -0,0 +1,125 @@
+OBJECT Codeunit 3041 DotNet_WebClient
+{
+  OBJECT-PROPERTIES
+  {
+    Date=;
+    Time=;
+    Version List=;
+  }
+  PROPERTIES
+  {
+    OnRun=BEGIN
+          END;
+
+  }
+  CODE
+  {
+    VAR
+      DotNetWebClient@1000 : DotNet "'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.System.Net.WebClient";
+      UrlSecurityErr@1001 : TextConst 'ENU=Security settings prevents download of this file.';
+
+    [External]
+    PROCEDURE Init@4();
+    BEGIN
+      DotNetWebClient := DotNetWebClient.WebClient;
+    END;
+
+    [External]
+    PROCEDURE SetCredentials@3(DotNet_Credential@1000 : Codeunit 3029);
+    VAR
+      DotNetCredential@1001 : DotNet "'System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.System.Net.NetworkCredential";
+    BEGIN
+      DotNet_Credential.GetCredential(DotNetCredential);
+      DotNetWebClient.Credentials := DotNetCredential;
+    END;
+
+    [External]
+    PROCEDURE DownloadData@5(Url@1000 : Text;VAR DotNet_MemoryStream@1001 : Codeunit 3009);
+    VAR
+      MemoryStream@1002 : DotNet "'mscorlib'.System.IO.MemoryStream";
+    BEGIN
+      VerifyUrl(Url);
+      DotNet_MemoryStream.SetMemoryStream(MemoryStream.MemoryStream(DotNetWebClient.DownloadData(Url)));
+    END;
+
+    LOCAL PROCEDURE DownloadString@7(Url@1000 : Text) : Text;
+    BEGIN
+      EXIT(DotNetWebClient.DownloadString(Url));
+    END;
+
+    LOCAL PROCEDURE VerifyUrl@6(Url@1000 : Text);
+    VAR
+      DotNet_Uri@1001 : Codeunit 3024;
+    BEGIN
+      DotNet_Uri.Init(Url);
```

Comment:
This method breaks the dotnet wrapper rules but as it is a security thing that's ok.

We have a bunch of similar functions in Codeunit 1299. In addition to http/https it verifies whether the url is well formed and valid. You could copy that functionality here.

Reply:
You still want me to use a codeunit that is not a wrapper and containes dotnet object?

We can do

WebRequestHelper.IsValidUri(Url);
IF NOT (WebRequestHelper.IsSecureHttpUrl(Url) OR WebRequestHelper.IsHttpUrl(Url)) THEN
  ERROR(UrlSecurityErr);

Reply:
No, the wrapper cannot take a dependency on a non-wrapper object. I was merely suggesting you copy paste the code so that the check is the same in both places.